ultravox/model/hubert.py

```diff
@@ -0,0 +1,119 @@
+from pathlib import Path
+
+import fairseq
```

Comment:
In general I'm skeptical of new requirements that are not standalone and `fairseq` is a quite heavy requirement with its own PyTorch requirements.

Can't we use the HuggingFace version instead? I don't see KMeans there, but that should be easy to add separately.

Reply:
Let me take another look at the HF version. It didn't look like it had the multilingual variant, and it was a different checkpoint, and given this evolved script we've been working with for a bit is operational (and I wasn't sure yet how you all are thinking about managing models/binaries, via HF or not), I figured I'd just send this for discussion. I agree it'd be simpler if we can just use that.

Reply:
After looking around the HuBERT parts we need are quite scattered on HF and IMO our best path is to host a copy of the data we want in either Ultravox or LMNT HF repo, let me know how you'd like to handle it (if Ultravox do I have access, should I point you to binaries to upload yourself, etc.).

Summary of findings:

- the canonical references for English and Multilingual (with kmeans bin as well) are not on HF, they are at:
  - https://github.com/facebookresearch/fairseq/tree/main/examples/hubert
  - https://github.com/facebookresearch/fairseq/blob/main/examples/speech_to_speech/docs/textless_s2st_real_data.md
- there are as you saw some facebook HF HuBERT repos but they only have the one base model file, only English, no km-bin (a non-Facebook HF repo does have the multilingual km-bin), and the model appears to be a different format
- As an elaboration for background as useful whether now or future, from discussion with @sharvil. we explicitly don't want to use the fine-tuned HuBERT model variants. Our intent here is to use HuBERT tokens as a proxy for phonemes and preserve more pronunciation/filler words/etc. The fine-tuned HuBERT models are a different token space and AIUI effectively overly-lossy for our needs.
- I hear you on `fairseq` being heavyweight, it looks like it's really just used to load the model so I'll see if I can port to straight torch and will follow up shortly. There is some logic in their utility routine but it doesn't seem too bananas.

Reply:
Thanks for checking this out.

If you feel like the lift is gonna be too much work, then we can just go with `fairseq` as well. My intention wasn't to block you.

Uploading to HF is also a good idea. Given the messy state of things, I think many people will appreciate the gesture.
If you do end up creating the straight PyTorch version, I think pushing to HF hub can be done as a follow up.

I don't have a strong preference as to where it should live, specially since you're the one doing the work. @juberti can probably give you access to our org as well.